src/hotspot/share/code/nmethod.cpp

```diff
@@ -1444,7 +1444,7 @@ void nmethod::unlink() {
   ClassUnloadingContext::context()->register_unlinked_nmethod(this);
 }
 
-void nmethod::purge(bool free_code_cache_data) {
+void nmethod::purge(bool free_code_cache_data, bool unregister_nmethod) {
```

Comment:
When will `free_code_cache_data` be `false`? I thought with the sort-before-free patch, freeing is always delayed after purging.

Reply:
The compiler eagerly allocates code blobs, so it can give them back if not needed any more (e.g. compiler thread stops due to dynamic adjustment, when shutting down the compilers, when the compiler fails to install an nmethod, ...).

Only when code unloading we always delay freeing (as this code shows).

Reply:
Do those take this path? AFAICS, they are subclasses of `RuntimeBlob`. I added `assert(!free_code_cache_data)` at the beginning of this method and it didn't crash right away.

Reply:
I ran tier 1-4 with no issues with `ShouldNotReachHere` sprinkled in, and did the same source code analysis, with the same result. Modified the change accordingly.